pkg/virt-controller/watch/vm.go

```diff
@@ -83,209 +126,427 @@ func (c *VMController) runWorker() {
+
+	// Only consider pods which belong to this vm
+	pods, err = c.filterMatchingPods(vm, pods)
+	if err != nil {
+		return err
+	}
+
+	if len(pods) > 1 {
+		logger.Reason(fmt.Errorf("More than one pod detected")).Error("That should not be possible, will not requeue")
 		return nil
+	} else if len(pods) == 1 && pods[0].Annotations[virtv1.OwnedByAnnotation] == "virt-handler" {
+		// Whenever we see that a matching pod is owned by virt-handler, we can mak handover as fulfilled
+		c.handoverExpectations.CreationObserved(key)
 	}
 
-	switch vm.Status.Phase {
-	case kubev1.VmPhaseUnset, kubev1.Pending:
-		// Schedule the VM
+	// If neddsSync is true (expectations fulfilled) we can make save assumptions if virt-handler or virt-controller owns the pod
+	needsSync := c.podExpectations.SatisfiedExpectations(key) && c.handoverExpectations.SatisfiedExpectations(key)
 
-		// Deep copy the object, so that we can safely manipulate it
-		vmCopy := vm.DeepCopy()
-		logger = log.Log.Object(vmCopy)
+	var syncErr error
+	if needsSync {
+		syncErr = c.sync(vm, pods)
+	}
+	return c.updateStatus(vm, pods, syncErr)
+}
 
-		// Check if there are already outdated VM Pods
-		pods, err := c.vmService.GetRunningVMPods(vmCopy)
-		if err != nil {
-			logger.Reason(err).Error("Fetching VM pods failed.")
-			return err
-		}
+func (c *VMController) updateStatus(vm *virtv1.VirtualMachine, pods []*k8sv1.Pod, syncErr error) error {
 
-		// If there are already pods, delete them before continuing ...
-		if len(pods.Items) > 0 {
-			logger.Error("VM Pods do already exist, will clean up before continuing.")
-			if err := c.vmService.DeleteVMPod(vmCopy); err != nil {
-				logger.Reason(err).Error("Deleting VM pods failed.")
-				return err
+	// Don't process states where the vm is clearly owned by virt-handler
+	if vm.IsRunning() || vm.IsScheduled() {
+		return nil
+	}
+
+	vmCopy := vm.DeepCopy()
+
+	if vm.IsUnprocessed() || vm.IsScheduling() {
+		if len(pods) > 0 {
+
+			containersAreReady := podIsReady(pods[0])
+
+			podIsTerminating := podIsDownOrGoingDown(pods[0])
+
+			// vm is still owned by the controller but pod is already handed over,
+			// so let's hand over the vm too
+			if c.isPodOwnedByHandler(pods[0]) {
+
+				vmCopy.Status.Interfaces = []virtv1.VirtualMachineNetworkInterface{
+					{
+						IP: pods[0].Status.PodIP,
+					},
+				}
+				vmCopy.Status.Phase = virtv1.Scheduled
+				if vmCopy.Labels == nil {
+					vmCopy.Labels = map[string]string{}
+				}
+				vmCopy.ObjectMeta.Labels[virtv1.NodeNameLabel] = pods[0].Spec.NodeName
+				vmCopy.Status.NodeName = pods[0].Spec.NodeName
+			} else if !podIsTerminating && !containersAreReady {
+				vmCopy.Status.Phase = virtv1.Scheduling
+			} else if podIsTerminating {
+				vmCopy.Status.Phase = virtv1.Failed
 			}
-			// the pod informer will reenqueue the key as a result of it being deleted.
-			return nil
+		} else if len(pods) == 0 && vm.IsScheduling() {
+			// someone other then the controller deleted the pod unexpectedly
```

Comment:
We only get to the updateStatus() function fi all expectations of pod creation/deletion/handover are fulfilled. If any expectation is still not satisfied we simply do nothing, so that should work.